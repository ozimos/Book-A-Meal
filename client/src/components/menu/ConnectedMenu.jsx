import React from 'react';
import ReactModal from 'react-modal';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  Accordion,
  AccordionItem,
  AccordionItemTitle,
  AccordionItemBody,
} from 'react-accessible-accordion';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import SearchInput, { createFilter } from 'react-search-input';
import MealDisplayCard from '../mealCard/MealDisplayCard';
import MealCardContainer from '../mealCard/MealCardContainer';
import Greeting from '../greeting/Greeting';
import { mealActions, menuActions } from '../../redux/actions';
import ConnectedNav from '../nav/ConnectedNav';
import '../../../public/styles/bookameal.scss';
import '../../../public/styles/modalOpenButton.scss';
import '../../../public/styles/search-input.css';
import '../../../public/styles/accordion.css';

ReactModal.setAppElement(document.getElementById('root'));
class Menu extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      searchTerm: '',
      menu: [],
      dbMenu: []
    };
    this.addToMenu = this.addToMenu.bind(this);
    this.postMenu = this.postMenu.bind(this);
  }
  static getDerivedStateFromProps(nextProps, prevState) {
    let incomingMenu = [];
    let prevMenu = [];
    if (prevState.dbMenu) {
      prevMenu = prevState.dbMenu.map(meal => meal.id);
    }
    if (nextProps.menu) {
      incomingMenu = nextProps.menu.map(meal => meal.id);
    }
    const incomingMenuSet = new Set(incomingMenu);
    const difference = prevMenu.filter(x => !incomingMenuSet.has(x));
    if (difference.length > 0) {
      const newMenu = [...nextProps.menu];
      newMenu
        .forEach((meal) => { delete meal.MealMenus; });
      return {
        dbMenu: nextProps.menu,
        menu: newMenu };
    }
    return null;
  }
  async componentDidMount() {
    this.props.dispatch(mealActions.getAllMeals());
    await this.props.dispatch(menuActions.getMenu());
    const newMenu = [...this.props.menu];
    newMenu
      .forEach((meal) => { delete meal.MealMenus; });
    // eslint-disable-next-line
     this.setState({ menu: newMenu, dbMenu: this.props.menu });
  }

  addToMenu(meal) {
    const inMenu = this.state.menu.some(elem => elem.id === meal.id);
    if (!inMenu) {
      this.setState(prevState => ({ menu: [...prevState.menu, meal] }));
      toast.success('Meal has been added to menu', { className: 'toaster' });
    } else {
      toast.error('Meal is already in menu', { className: 'toaster' });
    }
  }

  removeFromMenu = id =>
    this.setState(prevState =>
      ({ menu: prevState.menu.filter(elem => elem.id !== id) }));
  searchUpdated = (term) => {
    this.setState({ searchTerm: term });
  }
  postMenu() {
    const mealIdArray = this.state.menu.map(meal => meal.id);
    this.props.dispatch(menuActions.postMenu({ meals: mealIdArray }));
    toast.success('Menu for the day has been posted', { className: 'toaster' });
  }

  render() {
    const KEYS_TO_FILTERS = ['id', 'title', 'description', 'price'];

    const filteredMeals = this.props.meals
      .filter(createFilter(this.state.searchTerm, KEYS_TO_FILTERS));
    const { isCaterer, firstName } = this.props.user;
    const isMenuSet = this.state.menu.length !== 0;

    return (
      <div className="contain">
        <header className="header">
          <ConnectedNav />
        </header>
        <main>
          <ToastContainer autoClose={2000} />

          <Greeting isCaterer={isCaterer} firstName={firstName} />
          <Accordion accordion>
            <AccordionItem expanded>
              <AccordionItemTitle>
                <div className="title-element flexbox">
                  <h5>
                  Your Meals
                  </h5>
                  <div
                    className="accordion__arrow u-position-relative"
                    role="presentation"
                  />
                </div>
              </AccordionItemTitle>
              <AccordionItemBody>
                <SearchInput
                  className="search-input input-field"
                  onChange={this.searchUpdated}
                />
                <MealCardContainer
                  meals={filteredMeals}
                  MealCard={MealDisplayCard}
                  collection="Menu"
                  addToCollection={this.addToMenu}
                  addClass="scroll"
                />
              </AccordionItemBody>
            </AccordionItem>
            <AccordionItem>
              <AccordionItemTitle>
                <div className="title-element flexbox wrap">
                  <h5>
          Today&#39;s Menu
                  </h5>
                  <div
                    className="accordion__arrow u-position-relative"
                    role="presentation"
                  />
                </div>
              </AccordionItemTitle>
              <AccordionItemBody>
                <div className="title-element flexbox">
                  <button className="btn title-button" onClick={this.postMenu}>
                    <p>Post Menu</p>
                  </button>
                  <button
                    className="title-button btn"
                    onClick={() => this.setState({ menu: [] })
                  }
                  >
                    <p>Clear Menu</p>
                  </button>
                </div>

                {
                  isMenuSet ?
                    <MealCardContainer
                      meals={this.state.menu}
                      MealCard={MealDisplayCard}
                      collection="Menu"
                      removeFremoveFromCollectionromMenu={this.removeFromMenu}
                      addClass="scroll"
                    /> :
                    <div className="menu-message">
                  `You have not added meals to the menu.
                   Select a meal from the meal section
                   above to add meal to menu.`
                    </div>
                 }
              </AccordionItemBody>
            </AccordionItem>

          </Accordion>
        </main>
      </div>
    );
  }
}
Menu.defaultProps = {
  menu: []
};
Menu.propTypes = {
  dispatch: PropTypes.func.isRequired,
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  menu: PropTypes.arrayOf(PropTypes.object),
  user: PropTypes.shape({
    isCaterer: PropTypes.bool,
    firstName: PropTypes.string
  }).isRequired,
};
const mapStateToProps = state => ({
  mealError: state.mealsReducer.mealError,
  connecting: state.mealsReducer.connecting,
  deleted: state.mealsReducer.deleted,
  meals: state.mealsReducer.meals,
  menu: state.menuReducer.menu.Meals,
  authenticated: state.loginReducer.authenticated,
  user: state.loginReducer.user.data
});

export { Menu };
export default connect(mapStateToProps)(Menu);
