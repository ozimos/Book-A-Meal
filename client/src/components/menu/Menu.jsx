import React from 'react';
import ReactModal from 'react-modal';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  Accordion,
  AccordionItem,
  AccordionItemTitle,
  AccordionItemBody,
} from 'react-accessible-accordion';
import SearchInput, { createFilter } from 'react-search-input';
import MealCard2 from '../mealCard/MealCard2';
import MealCardContainer from '../mealCard/MealCardContainer';
import ConnectedGreeting from '../greeting/Greeting';
import { mealActions, menuActions } from '../../redux/actions';
import ConnectedNav from '../nav/Nav';
import '../../../public/styles/bookameal.scss';
import '../../../public/styles/modalOpenButton.scss';
import '../../../public/styles/search-input.css';
import '../../../public/styles/accordion.css';

ReactModal.setAppElement(document.getElementById('root'));
class Menu extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      searchTerm: '',
      menu: []
    };
  }
  componentDidMount() {
    this.props.dispatch(mealActions.getAllMeals());
  }

  addToMenu = (meal) => {
    const inMenu = this.state.menu.some(elem => elem.id === meal.id);
    if (!inMenu) { this.setState(prevState => ({ menu: [...prevState.menu, meal] })); }
  }
  removeFromMenu = id =>
    this.setState(prevState =>
      ({ menu: prevState.menu.filter(elem => elem.id !== id) }));
  searchUpdated = (term) => {
    this.setState({ searchTerm: term });
  }
  postMenu = () => {
    const mealIdArray = this.state.menu.map(meal => meal.id);
    this.props.dispatch(menuActions.postMenu({ meals: mealIdArray }));
  }

  render() {
    const KEYS_TO_FILTERS = ['id', 'title', 'description', 'price'];

    const filteredMeals = this.props.meals
      .filter(createFilter(this.state.searchTerm, KEYS_TO_FILTERS));
    const { isCaterer, firstName } = this.props.user;
    return (
      <div className="contain">
        <header className="header">
          <ConnectedNav />
        </header>
        <main>
          <ConnectedGreeting isCaterer={isCaterer} firstName={firstName} />
          <Accordion accordion={false}>
            <AccordionItem expanded>
              <AccordionItemTitle>
                <div className="title-element flexbox">
                  <h4>
                  Your Meals
                  </h4>
                  <div className="accordion__arrow u-position-relative" role="presentation" />
                </div>
              </AccordionItemTitle>
              <AccordionItemBody>
                <SearchInput className="search-input input-field" onChange={this.searchUpdated} />
                <MealCardContainer
                  meals={filteredMeals}
                  MealCard={MealCard2}
                  addToMenu={this.addToMenu}
                  addClass="scroll"
                />
              </AccordionItemBody>
            </AccordionItem>
            <AccordionItem>
              <AccordionItemTitle>
                <div className="title-element flexbox wrap">
                  <h3>
          Today&#39;s Menu
                  </h3>
                  <div className="accordion__arrow u-position-relative" role="presentation" />
                </div>
              </AccordionItemTitle>
              <AccordionItemBody>
                <div className="title-element flexbox">
                  <button className="btn title-button" onClick={this.postMenu}>
                    <p>Post Menu</p>
                  </button>
                  <button className="title-button btn" onClick={() => this.setState({ menu: [] })}>
                    <p>Clear Menu</p>
                  </button>
                </div>

                <MealCardContainer
                  meals={this.state.menu}
                  MealCard={MealCard2}
                  removeFromMenu={this.removeFromMenu}
                  addClass="scroll"
                />
              </AccordionItemBody>
            </AccordionItem>

          </Accordion>
        </main>
      </div>
    );
  }
}
Menu.propTypes = {
  dispatch: PropTypes.func.isRequired,
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  user: PropTypes.shape({
    isCaterer: PropTypes.bool,
    firstName: PropTypes.string
  }).isRequired,
};
const mapStateToProps = state => ({
  mealError: state.mealsReducer.mealError,
  connecting: state.mealsReducer.connecting,
  deleted: state.mealsReducer.deleted,
  meals: state.mealsReducer.meals,
  authenticated: state.loginReducer.authenticated,
  user: state.loginReducer.user.data
});

export { Menu };
export default connect(mapStateToProps)(Menu);
