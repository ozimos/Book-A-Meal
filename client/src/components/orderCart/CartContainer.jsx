import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { orderActions } from '../../redux/actions';


import '../../../public/styles/book_a_meal.css';
import '../../../public/styles/cart_layout.scss';


class CartContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      orderQuantity: {},
      prevPropsOrderId: this.props.orderId
    };
  }
  static getDerivedStateFromProps(nextProps, prevState) {
    const newOrderQuantity = {};
    if (nextProps.orderId !== prevState.prevPropsOrderId) {
      nextProps.order
        .forEach((elem) => { newOrderQuantity[elem.id] = elem.MealOrders.quantity; });
      return {
        prevPropsOrderId: nextProps.orderId,
        orderQuantity: newOrderQuantity };
    }
    return null;
  }
  componentDidMount() {
    if (this.props.orderId) {
      const newOrderQuantity = {};

      this.props.order
        .forEach((elem) => { newOrderQuantity[elem.id] = elem.MealOrders.quantity; });
      // eslint-disable-next-line
      this.setState({ orderQuantity: newOrderQuantity });

    }
  }
  setQuantity = (e, id) => {
    const inputValue = e.target.value;
    this.setState(prevState =>
      ({ orderQuantity: { ...prevState.orderQuantity, [id]: inputValue } }));
  };
 placeOrder = async () => {
   const actualOrder = this.props.order
     .map(meal => ({ id: meal.id,
       quantity: this.state.orderQuantity[meal.id] || 1 }));
   if (this.props.orderId) {
     await this.props
       .dispatch(orderActions.updateOrder({ meals: actualOrder }, this.props.orderId));
   } else {
     await this.props.dispatch(orderActions.postOrder({ meals: actualOrder }));
   }
   /* eslint no-alert: off */
   if (!this.props.orderError) {
     alert('Order has been created');
     this.props.clearCart();
   }
   if (this.props.orderError) {
     alert(this.props.orderError);
   }
 }

 render() {
   const { MealRow, order, ...rest } = this.props;
   const reducer = (total, meal) => total + (meal.price * (this.state.orderQuantity[meal.id] || 1));
   const calcTotal = () => this.props.order
     .reduce(reducer, 0);
   const total = calcTotal();
   return (
     <div className={rest.addClass ? `${rest.addClass}` : ''}>
       <h3>Order Cart</h3>
       <div className="table-responsive">
         <table className="table">
           <thead>
             <tr>
               <th>Remove Item</th>
               <th>Meal Title</th>
               <th>Price (&#x20a6;)</th>
               <th>Quantity</th>
               <th>Subtotal (&#x20a6;)</th>
             </tr>
           </thead>
           <tbody>
             {order.map(meal => (<MealRow
               key={meal.id}
               {...meal}
               setQuantity={this.setQuantity}
               quantity={(this.state.orderQuantity[meal.id] || '')}
               {...rest}
             />))
        }
           </tbody>
           <caption>
             <div className="row">
               <div className="col-3 mr-auto">&#9776; Your Order</div>
               <div className="col-2"> Total </div>
               <div className="col-2"> {total}</div>
               <div className="col2">
                 <button className="btn btn-cart" onClick={rest.clearCart}>
                   <p>Clear Cart</p>
                 </button>
               </div>
               <div className="col2">
                 <button className="btn btn-cart" onClick={this.placeOrder}>
                   {this.props.orderId ? <p>Modify Order</p> : <p>Place New Order</p>}
                 </button>
               </div>
             </div>
           </caption>
         </table>
       </div>
     </div>
   );
 }
}

CartContainer.defaultProps = {
  addClass: '',
  orderError: '',
  orderId: ''
};
CartContainer.propTypes = {
  order: PropTypes.arrayOf(PropTypes.object).isRequired,
  MealRow: PropTypes.func.isRequired,
  addClass: PropTypes.string,
  orderId: PropTypes.string,
  orderError: PropTypes.string,
  clearCart: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired,
};
const mapStateToProps = state => ({
  orderError: state.orderReducer.orderError,
  connecting: state.orderReducer.connecting,
});
export { CartContainer };
export default connect(mapStateToProps)(CartContainer);

