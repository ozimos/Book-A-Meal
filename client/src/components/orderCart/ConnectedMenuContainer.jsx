import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { menuActions } from '../../redux/actions';

import '../../../public/styles/cart_layout.scss';


class MenuContainer extends React.Component {
 placeOrder = async () => {

   if (this.props.postMenu) {
     await this.props.postMenu(this.props.menu);
   } else {
     await this.props.dispatch(menuActions.postMenu(this.props.menu));
   }

   if (!this.props.menuError) {
     const message = this.props.postMenu ? 'added to' : 'removed from';
     this.props.notify(`The selected meals have been ${message} the menu`);
     this.props.clearMenu();
   }

   if (this.props.menuError) {
     this.props.notify(this.props.menuError);
   }
 }

 render() {
   const { menu, ...rest } = this.props;

   return (
     <div className={rest.addClass ? `${rest.addClass}` : ''}>
       <div className="flexbox">
         <h5>Menu Editor</h5>
         <button className="btn title-button" onClick={rest.closeMenuModal}>
                &#10006;
         </button>
       </div>

       <div className="scroll">
         {menu.map(meal =>
             (
               <div key={meal.id} className="flexbox">
                 <span
                   className="text_left long_string"
                   key={meal.id}
                 >
                   {meal.title}
                 </span>
                 <button
                   className="btn title-button"
                   onClick={() => rest.removeFromMenu(meal.id)}
                 >
               Remove
                 </button>
               </div>
               ))
        }
       </div>

       <div className="flexbox">

         <button className="btn btn-cart" onClick={this.placeOrder}>
              Post meals to Menu
         </button>

         <button className="btn btn-cart" onClick={rest.clearMenu}>
                   Clear Menu
         </button>

       </div>

     </div>
   );
 }
}

MenuContainer.defaultProps = {
  addClass: '',
  menuError: '',
};

MenuContainer.propTypes = {
  menu: PropTypes.arrayOf(PropTypes.object).isRequired,
  addClass: PropTypes.string,
  menuError: PropTypes.string,
  clearMenu: PropTypes.func.isRequired,
  notify: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired,
  postMenu: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  menuError: state.menuReducer.menuError,
  connecting: state.menuReducer.connecting,
});

export { MenuContainer };
export default connect(mapStateToProps)(MenuContainer);

