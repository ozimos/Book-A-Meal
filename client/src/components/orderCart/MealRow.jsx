import React from 'react';
import PropTypes from 'prop-types';
import { hot } from 'react-hot-loader';

const MealRow = ({ id, title, price, quantity, setQuantity, removeFromCart }) => {
  const subTotal = price * (quantity || 1);
  return (
    <React.Fragment>
      <tr>
        <td>
          <button className="btn title-button icon" onClick={() => removeFromCart(id)}>
                &#10006;
          </button>
        </td>
        <td>{title}</td>
        <td>{price}</td>
        <td>
          <input
            className="formControl"
            type="number"
            placeholder="1"
            onChange={event => setQuantity(event, id)}
            value={quantity}
            style={{ width: '40px' }}
          />
        </td>
        <td>{subTotal}</td>
      </tr>
    </React.Fragment>
  );
};

MealRow.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  quantity: PropTypes.oneOfType(PropTypes.string, PropTypes.number).isRequired,
  removeFromCart: PropTypes.func.isRequired,
  setQuantity: PropTypes.func.isRequired,
};
export default hot(module)(MealRow);
