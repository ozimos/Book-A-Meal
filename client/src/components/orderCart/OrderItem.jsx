import React from 'react';
import PropTypes from 'prop-types';
import format from 'date-fns/format';
import differenceInMinutes from 'date-fns/difference_in_minutes';
import isToday from 'date-fns/is_today';
import '../../../public/styles/order_item.scss';

const OrderItem = ({ id, Meals, addOrder, position, updatedAt }) => {
  const details = Meals.map(meal => (
    <div key={meal.id} className="row">
      <div className="col s7 text_left long_string left_pad">{meal.title}</div>
      <div className="col s3 text_left left_pad">{meal.price}</div>
      <div className="col s2 text_left">{meal.MealOrders.quantity}</div>
    </div>));
  const hasEditMinutes = (parseInt(process.env.ORDER_EDIT_MINUTES, 10) || 15)
       - differenceInMinutes(new Date(), updatedAt) > 0;

  return (
    <div className="row order_item" style={{ borderBottom: '1px solid' }}>
      <div className="col s1 text_left">{position}</div>
      <div className="col s9">
        {details}
      </div>
      <div className="col s1 text_left">{format(updatedAt, 'h:m A')}</div>
      <div className="col s1 text_right">
        {
          isToday(updatedAt) && hasEditMinutes &&
          <button onClick={() => addOrder(id)} className="btn title-button">Edit</button>
        }
      </div>
    </div>
  );
};

OrderItem.propTypes = {
  id: PropTypes.string.isRequired,
  updatedAt: PropTypes.string.isRequired,
  position: PropTypes.number.isRequired,
  Meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  addOrder: PropTypes.func.isRequired
};
export default OrderItem;

