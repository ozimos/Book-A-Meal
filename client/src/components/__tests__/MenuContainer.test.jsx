/* global React:false, shallow:false toJson:false */

import { MenuContainer } from '../orderCart/ConnectedMenuContainer';
import menu from '../mocks/menuDataMock';

describe('MenuContainer Component', () => {

  it('should render without throwing an error', () => {
    expect(shallow(<MenuContainer
      menu={menu.data.rows}
      clearMenu={jest.fn()}
      dispatch={jest.fn()}
      postMenu={jest.fn()}
    />)
      .exists(<table className="table" />)).toBe(true);
  });
  it('renders correctly', () => {
    const wrapper = shallow(<MenuContainer
      menu={menu.data.rows}
      clearMenu={jest.fn()}
      dispatch={jest.fn()}
      postMenu={jest.fn()}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

