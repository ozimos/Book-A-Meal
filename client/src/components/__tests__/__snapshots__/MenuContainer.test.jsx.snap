// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuContainer Component renders correctly 1`] = `
<div
  className=""
>
  <div
    className="flexbox"
  >
    <h5>
      Menu Editor
    </h5>
    <button
      className="btn title-button"
    >
      âœ–
    </button>
  </div>
  <div
    className="scroll"
  >
    <div
      className="flexbox"
    >
      <span
        className="text_left long_string"
      />
      <button
        className="btn title-button"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </div>
  <div
    className="flexbox"
  >
    <button
      className="btn btn-cart"
      onClick={[Function]}
    >
      Post meals to Menu
    </button>
    <button
      className="btn btn-cart"
      onClick={[MockFunction]}
    >
      Clear Menu
    </button>
  </div>
</div>
`;
